rclc_parameter/src/rclc_parameter/parameter_server.c

```diff
@@ -0,0 +1,618 @@
+  rclc_parameter_server_t * param_server = (rclc_parameter_server_t *) parameter_server;
+
+  if (request->parameters.size > response->results.capacity) {
+    response->results.size = 0;
+    return;
+  }
+
+  response->results.size = request->parameters.size;
+
+  for (size_t i = 0; i < response->results.size; i++) {
+    rosidl_runtime_c__String * message =
+      (rosidl_runtime_c__String *) &response->results.data[i].reason;
+    Parameter * parameter = rclc_parameter_search(
+      &param_server->parameter_list,
+      request->parameters.data[i].name.data);
+    rcl_ret_t ret = RCL_RET_OK;
+
+    if (parameter != NULL) {
+      response->results.data[i].successful = true;
+
+      if (parameter->value.type != request->parameters.data[i].value.type) {
+        rclc_parameter_set_string(message, "Type mismatch");
+        response->results.data[i].successful = false;
+        continue;
+      }
+
+      switch (request->parameters.data[i].value.type) {
+        case RCLC_PARAMETER_NOT_SET:
+          rclc_parameter_set_string(message, "Type not set");
+          response->results.data[i].successful = false;
+          break;
+
+        case RCLC_PARAMETER_BOOL:
+          ret =
+            rclc_parameter_set_bool(
+            parameter_server, parameter->name.data,
+            request->parameters.data[i].value.bool_value);
+          break;
+        case RCLC_PARAMETER_INT:
+          ret =
+            rclc_parameter_set_int(
+            parameter_server, parameter->name.data,
+            request->parameters.data[i].value.integer_value);
+          break;
+        case RCLC_PARAMETER_DOUBLE:
+          ret =
+            rclc_parameter_set_double(
+            parameter_server, parameter->name.data,
+            request->parameters.data[i].value.double_value);
+          break;
+
+        default:
+          rclc_parameter_set_string(message, "Type not supported");
+          response->results.data[i].successful = false;
+          break;
+      }
+
+      if (ret == RCL_RET_INVALID_ARGUMENT) {
+        rclc_parameter_set_string(message, "Set parameter error");
+        response->results.data[i].successful = false;
+      }
+    } else {
+      rclc_parameter_set_string(message, "Parameter not found");
+      response->results.data[i].successful = false;
+    }
+  }
+}
+
+rcl_ret_t rclc_parameter_server_init_default(
+  rclc_parameter_server_t * parameter_server,
+  rcl_node_t * node)
+{
+  rclc_parameter_options_t opts = {
+    .notify_changed_over_dds = true,
+    .max_params = 4
+  };
+  return rclc_parameter_server_init_with_option(parameter_server, node, &opts);
+}
+
+rcl_ret_t rclc_parameter_server_init_with_option(
+  rclc_parameter_server_t * parameter_server,
+  rcl_node_t * node,
+  rclc_parameter_options_t * options)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(parameter_server, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(node, RCL_RET_INVALID_ARGUMENT);
+
```

Comment:
`options` not checked for NULL pointer. 

check entire source code, if you test for NULL pointer for all function arguments.

Reply:
Done f111684c6414527f89c34358cdc0fe0460fc61e1